### PR TITLE
Fix/empty string datetimelike conversion/issue 36550

### LINKED ISSUE #36550: BUG: Categorical series `.isin({pd.NaT, ...})` gives all-False
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import pandas as pd
pd.Series([pd.NaT, ''], dtype='category').isin({pd.NaT, ''})  # [True, True]
pd.Series([''], dtype='category').isin({''})  # [True] because '' is in series
pd.Series([''], dtype='category').isin({None, ''})  # [True] because '' is in series
pd.Series([''], dtype='category').isin({pd.NaT, ''})  # [False] even though '' is in series
```

#### Problem description

When searching for empty values in a categorical full of empty values, searching for `pd.NaT` will prevent other terms from being found.

(Perhaps that's because Pandas casts all input values to Timedelta?)

This behavior is unexpected. It makes me feel like I must be crazy. If I'm not meant to search for `pd.NaT`, I'd feel very happy with an error message when I try this.

#### Expected Output

```python
pd.Series([''], dtype='category').isin({pd.NaT, ''})  # [True] because '' is in series
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 2a7d3326dee660824a8433ffd01065f8ac37f7d6
python           : 3.8.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.8.4-200.fc32.x86_64
Version          : #1 SMP Wed Aug 26 22:28:08 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.1.2
numpy            : 1.19.2
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 45.2.0
Cython           : None
pytest           : 5.4.3
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : None
pandas_datareader: None
bs4              : 4.9.1
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 0.16.0
pytables         : None
pyxlsb           : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

</details>

### PR DESCRIPTION
**Pull request by nrebena**
- [x] xref #36550
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

# Empty string conversion

The actual comportment of Timedelta, Timestamp and to_datetime is to return NaT.
```python
import pandas as pd
pd.to_datetime("", errors="raise")
# NaT
pd.to_datetime("", errors="coerce")
# NaT
pd.to_datetime("", errors="ignore")
# NaT
pd.Timedelta("")
# NaT
pd.Timestamp("")
# NaT
```

I would propose to raise a ValueError instead, same as for string that are not datetimelike, as follow:

 ```python
import pandas as pd
import pytest

with pytest.raises(ValueError):
    pd.to_datetime("", errors="raise")
pd.to_datetime("", errors="coerce")
# NaT
pd.to_datetime("", errors="ignore")
# ''

with pytest.raises(ValueError):
    pd.Timedelta("")
with pytest.raises(ValueError):
    pd.Timestamp("")
```

# Tests

I added the relevant test for the expected comportment.
I still will have to fix some test that rely on the current behaviors.

# Benchmark

I ran the following command
`asv continuous -E virtualenv -f 1.1 upstream/master HEAD -b "^tslibs.(timestamp|timedelta|tslib)"`
<details>

```
       before           after         ratio
     [8f6ec1e8]       [4393cc9e]
     <master>         <fix/empty-string-datetimelike-conversion/issue-36550>
+         173±1ns         202±10ns     1.17  tslibs.timestamp.TimestampProperties.time_is_leap_year(tzlocal(), 'B')
+         162±2ns          188±4ns     1.16  tslibs.timestamp.TimestampProperties.time_is_quarter_start(tzlocal(), None)
+         178±1ns         205±20ns     1.15  tslibs.timestamp.TimestampProperties.time_dayofweek(tzfile('/usr/share/zoneinfo/US/Central'), 'B')
+         163±2ns         186±10ns     1.15  tslibs.timestamp.TimestampProperties.time_is_year_end(tzlocal(), None)
+         162±1ns         185±10ns     1.15  tslibs.timestamp.TimestampProperties.time_is_year_end(tzfile('/usr/share/zoneinfo/US/Central'), None)
+         162±2ns         185±10ns     1.14  tslibs.timestamp.TimestampProperties.time_is_year_end(<UTC>, None)
+         169±2ns         193±10ns     1.14  tslibs.timestamp.TimestampProperties.time_is_quarter_end(tzlocal(), None)
+         164±2ns          186±4ns     1.14  tslibs.timestamp.TimestampProperties.time_is_quarter_start(tzutc(), None)
+     4.60±0.03μs       5.16±0.3μs     1.12  tslibs.timestamp.TimestampProperties.time_is_year_end(datetime.timezone(datetime.timedelta(seconds=3600)), 'B')
+         147±1ns          163±4ns     1.11  tslibs.timestamp.TimestampProperties.time_microsecond(<DstTzInfo 'Europe/Amsterdam' LMT+0:20:00 STD>, 'B')
+         170±2ns          188±7ns     1.11  tslibs.timestamp.TimestampProperties.time_is_quarter_end(tzfile('/usr/share/zoneinfo/US/Central'), None)
+       179±0.7ns         197±10ns     1.10  tslibs.timestamp.TimestampProperties.time_dayofweek(datetime.timezone(datetime.timedelta(seconds=3600)), None)
+       174±0.3ns         191±10ns     1.10  tslibs.timestamp.TimestampProperties.time_is_leap_year(<DstTzInfo 'Europe/Amsterdam' LMT+0:20:00 STD>, 'B')

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```
</details>

### CODE REVIEW

**Comment by jreback** [comment]
why this change? likely to break test parsing generally and would have to see a lot of benefit

**Comment by nrebena** [comment]
Considering empty string equivalent to NaT as a huge influence on all method that try to cast data into datetimelike (happen in groupby, merge, join, equal, probably others methods)
Globaly, I think we should refrain from **implicit** conversion of string to datetimelike, and this PR goes one step in this direction.
An other issue related to mixing string and datetimelike can be found here #36545 

For information the broken test were the following:
```
pandas/tests/io/parser/test_parse_dates.py::test_parse_dates_empty_string[c_high]
pandas/tests/io/parser/test_parse_dates.py::test_parse_dates_empty_string[c_low]
pandas/tests/io/parser/test_parse_dates.py::test_parse_dates_empty_string[python]
pandas/tests/plotting/test_datetimelike.py::TestTSPlot::test_ts_plot_with_tz['UTC']
pandas/tests/tools/test_to_datetime.py::TestToDatetimeMisc::test_to_datetime_types[True]
pandas/tests/tools/test_to_datetime.py::TestToDatetimeMisc::test_to_datetime_types[False]
pandas/tests/tools/test_to_timedelta.py::TestTimedeltas::test_to_timedelta
```

If empty string should be always considered as NaT, then I'll close this PR.

File: pandas/tests/scalar/test_nat.py [hunk 1]
```diff
@@ -104,8 +104,13 @@ def test_identity(klass, value):
 @pytest.mark.parametrize("klass", [Timestamp, Timedelta, Period])
 @pytest.mark.parametrize("value", ["", "nat", "NAT", None, np.nan])
 def test_equality(klass, value):
```
Comment (jreback) [inline pandas/tests/scalar/test_nat.py:106]: I would instead remove the empty and put this in a new test for that case (and assert the error)

these are duplicating the constructor tests so should have either / or (i think they are fine here)

**Review by jreback — CHANGES_REQUESTED** [review]
does this affect `read_csv`? do we have tests for this?

this needs a fairly big section i the whatsnew as this can break existing code

**Comment by jreback** [comment]
does this actually close the referenced issue?

**Comment by jreback** [comment]
can you merge master

**Comment by arw2019** [comment]
Closing for now. @nrebena let us know if you'd like to start this up again and we'll reopen

**Pull request closed without merging**